### DISCUSSION TITLE
Converting a list of GO terms to a directed acyclic graph

### DISCUSSION THREAD

**Discussion by Tim-Kirkwood** [geneontology/helpdesk#506] [Q&A] [answered]
Hello!  

I am making an app using Python and, as part of this, I would like to convert a collection of GO terms, associated with a single protein, to a DAG.  As an example of such a list:

`cardiac muscle cell development [GO:0055013]; cardiac muscle contraction [GO:0060048]; cardiac muscle hypertrophy [GO:0003300]; cardiac muscle tissue morphogenesis [GO:0055008]; cardiac myofibril assembly [GO:0055003]; detection of muscle stretch [GO:0035995]; mitotic chromosome condensation [GO:0007076]; muscle contraction [GO:0006936]; positive regulation of gene expression [GO:0010628]; positive regulation of protein secretion [GO:0050714]; protein kinase A signaling [GO:0010737]; response to calcium ion [GO:0051592]; sarcomere organization [GO:0045214]; sarcomerogenesis [GO:0048769]; skeletal muscle myosin thick filament assembly [GO:0030241]; skeletal muscle thin filament assembly [GO:0030240]; striated muscle contraction [GO:0006941]
`

In the DAG, I expect the `response to calcium ion [GO:0051592]` node would have an edge pointing to all of the muscle contraction terms (if we assume that the edges are from parent to child terms, child to parent is also fine).

I'm happy making the graph, but I'm not sure how to get the edges and their directions.  Also, I would like to ignore any terms that are not in the list, unless they are the only connection joining one of the list terms to another.  Ideally i'd like to get this via the [API](https://geneontology.org/docs/tools-guide/), but I will also be looking into [goatools](https://github.com/tanghaibao/goatools).  

If anyone else can offer any tips please let me know!

**Comment by sierra-moxon** [comment]
@sierra-moxon - tagging myself so that I don't lose track of this issue.

**Comment by balhoff** [comment, marked as answer]
It's not a web API, but the ROBOT tool has good support for this (whether you get a DAG is a bit dependent on the input though). The [extract command](https://robot.obolibrary.org/extract) provides a few different strategies. I think either `MIREOT` or `subset` might work best.

**Reply by Tim-Kirkwood** [reply]
Thanks very much, this is interesting - subset looks like a good fit!  I will try it out tomorrow and see how it goes.  

Out of interest, I had another look at the [go-basic.obo](https://current.geneontology.org/ontology/go-basic.obo) file - can I confirm this also has the information I am looking for?  For example, from the below snippet in the obo file, GO:0000006 is a child of GO:0005385:

```
[Term]
id: GO:0000006
name: high-affinity zinc transmembrane transporter activity
namespace: molecular_function
def: "Enables the transfer of zinc ions (Zn2+) from one side of a membrane to the other, probably powered by proton motive force. In high-affinity transport the transporter is able to bind the solute even if it is only present at very low concentrations." [TC:2.A.5.1.1]
synonym: "high affinity zinc uptake transmembrane transporter activity" EXACT []
synonym: "high-affinity zinc uptake transmembrane transporter activity" RELATED []
is_a: GO:0005385 ! zinc ion transmembrane transporter activity
```

**Reply by balhoff** [reply]
@Tim-Kirkwood go-basic is supposed to be a DAG, so it is probably a good starting point for you.
